lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetSchemaConverter.java

```diff
@@ -71,56 +72,82 @@
     public static final boolean HIVE_PARQUET_USE_LEGACY_DECIMAL_ENCODING = true;
     public static final boolean HIVE_PARQUET_USE_INT96_TIMESTAMP_ENCODING = true;
 
-    private Map<List<String>, Type> primitiveTypes = new HashMap<>();
-    private final boolean useLegacyDecimalEncoding;
-    private final boolean useInt96TimestampEncoding;
+    private final Map<List<String>, Type> primitiveTypes;
     private final MessageType messageType;
 
     public ParquetSchemaConverter(List<Type> types, List<String> columnNames, boolean useLegacyDecimalEncoding, boolean useInt96TimestampEncoding)
     {
         requireNonNull(types, "types is null");
         requireNonNull(columnNames, "columnNames is null");
         checkArgument(types.size() == columnNames.size(), "types size not equals to columnNames size");
-        this.useLegacyDecimalEncoding = useLegacyDecimalEncoding;
-        this.useInt96TimestampEncoding = useInt96TimestampEncoding;
-        this.messageType = convert(types, columnNames);
+        ImmutableMap.Builder<List<String>, Type> primitiveTypesBuilder = ImmutableMap.builder();
+        messageType = convert(types, columnNames, useLegacyDecimalEncoding, useInt96TimestampEncoding, primitiveTypesBuilder::put);
+        primitiveTypes = primitiveTypesBuilder.buildOrThrow();
     }
 
-    private MessageType convert(List<Type> types, List<String> columnNames)
+    public Map<List<String>, Type> getPrimitiveTypes()
```

Comment:
we could group `primitiveTypes` and `messageType` under one `record` and transform the `ParquetSchemaConverter` into an utility class responsible of creating `ParquetSchemaTypeMapping` types.

Reply:
let's not do that yet; the incurred list manipulations could be a perf loss